doc/manual/metaprogramming.rst

```diff
@@ -872,3 +872,203 @@ entirely in Julia. You can read their source and see precisely what they
 do — and all they do is construct expression objects to be inserted into
 your program's syntax tree.
 
+Staged functions
+----------------
+
+*Staged functions* play a similar role as macros, but at a later stage
+between parsing and run-time. Staged functions give the capability to
+generate specialized code depending on the types of their arguments.
+While macros work with expressions at parsing-time and cannot access the
+types of their inputs, a staged function gets expanded at a time when
+the types of the arguments are known, but the function is not yet compiled.
+
+Depending on the types of the arguments, a staged function returns a quoted
+expression which then forms the function body of the specialized function.
+Thus, staged functions provide a flexible framework to move work from
+run-time to compile-time.
+
+When defining staged functions, there are three main differences to
+ordinary functions:
+
+1. You use the keyword ``stagedfunction`` instead of ``function``
+
+2. In the body of the ``stagedfunction`` you only have access to the
+   *types* of the arguments, not their values.
+
+3. Instead of calculating something or performing some action, you return
+   from the staged function a *quoted expression* which, when evaluated,
+   does what you want.
+
+It's easiest to illustrate this with an example. We can declare a staged
+function ``foo`` as
+
+.. doctest::
+
+    julia> stagedfunction foo(x)
+               println(x)
+               :(x*x)
```

Comment:
i think it is best practice to write this as `return :(x*x)` in the manual, for enhanced clarify